### DISCUSSION TITLE
Transaction Conditional Errors

### DISCUSSION THREAD

**Discussion by garretcharp** [sensedeep/dynamodb-onetable#135] [Ideas]
Currently, whenever a transaction fails at least for unique items (haven't played with transactions without unique yet) you get a very basic error that does not really explain exactly why it failed just that something failed because of a condition check.

Example error currently: `dynamo: Cannot create "User", an item of the same name already exists.`

A nicer error would explain the condition check fails for you telling you what item failed. Let say for example you have a User model with the following unique attribute: Username, and Email and all the possible error cases for them:

### Username already exists:
Current: `dynamo: Cannot create "User", an item of the same name already exists.`
Wanted:
```js
{
    message: 'dynamo: Cannot create unique attribute "username" for "User".',
    details: [
        {
            message: 'dynamo: Cannot create unique attribute "username" for "User".',
            code: 'ConditionalCheckFailed',
            item: { pk: '_unique#User#username#test', type: '_Unique' }
        }
    ]
}
```

### Email already exists:
Current: `dynamo: Cannot create "User", an item of the same name already exists.`
Wanted:
```js
{
    message: 'dynamo: Cannot create unique attribute "email" for "User".',
    details: [
        {
            message: 'dynamo: Cannot create unique attribute "email" for "User".',
            code: 'ConditionalCheckFailed',
            item: { pk: '_unique#User#email#test@example.com', type: '_Unique' }
        }
    ]
}
```

### Username & Email already exists:
Current: `dynamo: Cannot create "User", an item of the same name already exists.`
Wanted:
```js
{
    message: 'dynamo: Cannot create unique attribute "username" for "User". dynamo: Cannot create unique attribute "email" for "User".',
    details: [
        {
            message: 'dynamo: Cannot create unique attribute "username" for "User".',
            code: 'ConditionalCheckFailed',
            item: { pk: '_unique#User#username#test', type: '_Unique' }
        },
        {
            message: 'dynamo: Cannot create unique attribute "email" for "User".',
            code: 'ConditionalCheckFailed',
            item: { pk: '_unique#User#email#test@example.com', type: '_Unique' }
        }
    ]
}
```

### The main model failed to create (not unique attributes)
Current: `dynamo: Cannot create "User", an item of the same name already exists.`
Wanted:
```js
{
    message: 'dynamo: Conditional create failed for "User".',
    details: [
        {
            message: 'dynamo: Conditional create failed for "User".',
            code: 'ConditionalCheckFailed',
            item: { pk: 'user#MY-USER-ID', username: 'test', email: 'test@example.com', _type: 'User' }
        }
    ]
}
```

**Comment by AndrewCathcart** [comment]
+1 for this. I also think there's currently some inconsistencies around what Error is returned depending on the type of operation, depending on whether it is an update or a put, which makes error handling a bit unintuitive with the library.

```typescript
            catch (err) {
                if (params.throw === false) {
                    result = {};
                }
                else if (err.code == 'ConditionalCheckFailedException' && op == 'put') {
                    //  Not a hard error -- typically part of normal operation
                    this.log.info(`Conditional check failed "${op}" on "${model}"`, { err, trace });
                    throw new Error(`Conditional create failed for "${model}`); // we lose the type of error here, and just get a basic message back (can't check e.g. for error.name being 'ConditionalCheckFailedException')
                }
                else {
                    result = result || {};
                    result.Error = 1;
                    trace.err = err;
                    if (params.log != false) {
                        this.log.error(`OneTable exception in "${op}" on "${model}"`, { err, trace });
                    }
                    throw err;
                }
            }
```

**Reply by mobsense** [reply]
Agree. We'll create a OneError class and then progressively unite all error handling via that. See more details below

**Comment by mobsense** [comment]
Thanks for the additional push. This is next on our list.

**Comment by mobsense** [comment]
Working on this now.

FYI: the thrown error object does include "details" already, but can be improved. Also details is not rendered in the toString.

**Comment by mobsense** [comment]
Some comments and progress details

- Dynamo doesn't give us the name of the condition that failed. So we can only present the set of unique attributes and can't pinpoint the exact one that failed in a transaction.

But I think we can give some pretty detailed errors. Errors will throw a OneError instance which is a subclass of error.

OneError Object Properties

* message
* code
* context
* context.properties
* context.transaction
* context.validations -- map for validation errors. Each failing property is listed. 
* stack

```
  message: Cannot create unqiue attributes "email" for "User", an item of the same name already exists.
  code: UniqueError
  context: {
      "properties": {
          "pk": "User#",
          "sk": "User#1635319324769",
          "gs1pk": "User#",
          "gs1sk": "User#99d2a1e483ddf6d19dc36f6da3adc6cd",
          "id": "99d2a1e483ddf6d19dc36f6da3adc6cd",
          "email": "coy@acme.com",
          "date": 1635319324769,
          "_type": "User"
      },
      "transaction": {
          "TransactItems": [
              {
                  "Put": {
                      "ConditionExpression": "(attribute_not_exists(pk)) and (attribute_not_exists(sk))",
                      "TableName": "DebugTable",
                      "Item": {
                          "pk": {
                              "S": "_unique#User#email#coy@acme.com"
                          },
                          "sk": {
                              "S": "_unique#"
                          },
                          "_type": {
                              "S": "_Unique"
                          }
                      },
                      "ReturnValues": "NONE"
                  }
              },
              {
                  "Put": {
                      "ConditionExpression": "(attribute_not_exists(pk)) and (attribute_not_exists(sk))",
                      "TableName": "DebugTable",
                      "Item": {
                          "pk": {
                              "S": "User#"
                          },
                          "sk": {
                              "S": "User#1635319324769"
                          },
                          "gs1pk": {
                              "S": "User#"
                          },
                          "gs1sk": {
                              "S": "User#99d2a1e483ddf6d19dc36f6da3adc6cd"
                          },
                          "id": {
                              "S": "99d2a1e483ddf6d19dc36f6da3adc6cd"
                          },
                          "email": {
                              "S": "coy@acme.com"
                          },
                          "date": {
                              "N": "1635319324769"
                          },
                          "_type": {
                              "S": "User"
                          }
                      },
                      "ReturnValues": "NONE"
                  }
              }
          ]
      },
      "code": "Unique"
  }
  stack: OneError: Cannot create unqiue attributes "email" for "User", an item of the same name already exists.
      at Model.createUnique (/Users/mob/dev/onetable/src/Model.js:514:23)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Model.create (/Users/mob/dev/onetable/src/Model.js:472:22)
      at Object.<anonymous> (/Users/mob/dev/onetable/test/debug.ts:69:9)
```

**Reply by AndrewCathcart** [reply]
Might be being a tad ignorant here but how are you getting the information "an item of the same name already exists." in OneError? 

I can understand how it might work for a Put, but If for example you're doing an Update with an ConditionExpression that isn't just `(attribute_not_exists(pk)) and (attribute_not_exists(sk))`, you could run into an error where e.g. the item doesn't exist in the table, or the extra condition in your where clause is not met (hope i'm being clear here).

As an example;
```
await EntitlementService.update(
      {
        emailAddress, // could return an error if an item with the specified pk and sk don't exist
        holidayPeriod: startPeriod,
      },
      {
        add: {
          remaining: -startPeriodCost,
        },
        where: '${remaining} >= @{startPeriodCost}', // could return an error
        substitutions: {
          startPeriodCost: startPeriodCost,
        },
        transaction,
      },
    );
```

**Reply by mobsense** [reply]
Let me check that out to see whether we can discriminate between the two.

**Comment by AndrewCathcart** [comment]
Appreciate the work so far @mobsense! Might be veering slightly off topic here, but I wonder if it would be useful for the users of your library to have some examples of dealing with error handling in the documentation? 

I'm currently torn because the errors that dynamo spits out when a transaction fails are not exactly obvious, which has caused me to resort to doing multiple dynamo calls to validate that e.g. the item exists, or otherwise doing questionable checks on the error name. Removed a bunch of code for conciseness (error thrown, logging etc), but just as a reference this is how I'm currently handling errors for e.g. update;

```typescript
let entitlement;
  try {
    entitlement = await EntitlementService.update({
    // ...
    });
  } catch (error) {
    if (error instanceof Error && error.name.includes('ConditionalCheckFailedException')) {
    // throw e.g. a 404, as the pk and sk ConditionExpression has probably failed (we don't have a where clause in the update)
    } else {
    // throw something like a 500 unknown error occurred (dynamo died / some connection issue)
    }
  }
```

Happy to create a separate discussion if needs be - I think it would be an interesting chat about handling errors / coding style. Figured I'd start here as we are discussing error handling / transactions / error messages.

**Comment by mobsense** [comment]
Examples would be great. We could add some to the README and create a sample that demonstrated the different techniques.

Having OneError in place gives us the structure and information to be able to differentiate errors. (Except for transactions where DynamoDB just doesn't seem to give enough info).

OneError will have a code which makes switching a little easier.

I'll try to push out this OneError update soon, and then we can easily iterate in specific places where the error is ambiguous or unclear.

**Comment by mobsense** [comment]
GitHub updated with OneError updates

**Comment by mobsense** [comment]
This has now been released in 2.0.3.

Please point out where the error messages are still unclear, insufficient or inconsistent. We now have a basis to address those well.
